torchx/tracker/mlflow.py

```diff
@@ -0,0 +1,337 @@
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import dataclasses
+import os
+import socket
+from getpass import getuser
+from logging import getLogger, Logger
+from pathlib import Path
+from tempfile import gettempdir
+from typing import Any, Iterable, Mapping, Optional, Sequence
+
+import mlflow
+from mlflow.entities import Run
+
+from torchx.distributed import on_rank0_first
+from torchx.runner.config import get_configs
+from torchx.tracker.api import Lineage, TrackerArtifact, TrackerBase, TrackerSource
+
+log: Logger = getLogger(__name__)
+TAG_ARTIFACT_MD_PREFIX = "torchx.artifact.metadata"
+
+
+class MLflowTracker(TrackerBase):
+    """
+    An implementation of a ``Tracker`` that uses mlflow as the backend.
+    Don't forget to call the ``close()`` method for orderly shutdown.
+    This ensures that the run state in mlflow is properly marked as ``FINISHED``,
+    otherwise the run will remain in ``UNFINISHED`` status.
+
+    .. important::
+        TorchX's run_id is used as mlflow's run_name! The run_id in TorchX
+        is the job name. The job name in TorchX is made unique by adding
+        a short random hash to the user-provided job name prefix. This is
+        done because certain job schedulers supported by TorchX requires
+        that the job name on the submitted job definition is globally unique
+        (rather than the scheduler returning a unique job id as the return result
+        of the job submission API).
+
+    .. warning::
+        APIs on this class may only be called with the same ``run_name`.
+        Typically the user does not have to worry about manually setting
+        the run_name as it is picked up by default from the environment variable
+        ``TORCHX_APP_NAME``.
+
+    """
+
+    def __init__(
+        self,
+        experiment_name: Optional[str] = None,
+        tracking_uri: str = f"file://{Path(gettempdir()) / 'torchx' / 'mlruns'}",
+        artifact_location: Optional[str] = None,
+    ):
+        if experiment_name is None:
+            experiment_name = self.default_experiment_name()
+
+        self.tracking_uri = tracking_uri
+        mlflow.set_tracking_uri(tracking_uri)
+        log.info(
+            f"MLflow tracking_uri={tracking_uri}, artifact_location={artifact_location}"
+        )
+        with on_rank0_first():
+            self.experiment = mlflow.get_experiment_by_name(experiment_name)
+            if self.experiment is None:
+                experiment_id = mlflow.create_experiment(
+                    name=experiment_name,
+                    artifact_location=artifact_location,
+                )
+                self.experiment = mlflow.get_experiment(experiment_id)
+                log.info(
+                    f"Created new experiment `{experiment_name}` (id={experiment_id})"
+                )
+            else:
```

Comment:
do we need any safety checks to avoid accidental clobbering? I suppose with the random ID it's fine

Reply:
yeah torchx job names are unique enough. The `on_rank0_first` guard seems to be enough - based on the distributed runs we've been running with this tracker for the past month.